File: codility/cpp/codility_lesson7_4.cpp
// you can use includes, for example:
// #include <algorithm>
#include <stack>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(vector<int> &H) {
	// write your code in C++14 (g++ 6.2.0)
	int N = H.size();
	int res = 0;
	stack<int> s;

	for (int i = 0; i < N; i++)
	{
		while (!s.empty() && H[i] < s.top())
			s.pop();

		if (s.empty() || H[i] > s.top())
		{
			res++;
			s.push(H[i]);
		}
	}

	return res;
}

File: codility/python/codility_lesson7_4.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(H):
	# write your code in Python 3.6
	ans = 0
	stack = []
	for h in H:
		if len(stack) == 0 or stack[-1] < h:
			stack.append(h)
			ans += 1
		elif stack[-1] > h:
			while len(stack) != 0 and stack[-1] > h:
				stack.pop()
			if len(stack) == 0 or stack[-1] != h:
				stack.append(h)
				ans += 1
	return ans
